File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.py
def solution(coin, cards):
    n = len(cards)
    target = n+1
    initial = n//3
    
    isFirst = [False for _ in range(n+1)]
    visited = [False for _ in range(n+1)]
    canBuy = 0
    canPlay = 1
    for i in range(initial):
        card = cards[i]
        pair = target - card
        isFirst[card] = True
        
        if(isFirst[pair]):
            canPlay += 1 
    
    round = 0
    while(canPlay > 0):
        canPlay -= 1
        
        idx = initial + (round * 2)
        if(idx >= n):
            break
            
        # 첫번째 카드
        card1 = cards[idx]
        card2 = cards[idx+1]
        visited[card1] = True
        visited[card2] = True
        
        pair1 = target - card1
        pair2 = target - card2
        if(coin > 0 and isFirst[pair1]):
            canPlay += 1
            coin -= 1
        if(coin > 0 and isFirst[pair2]):
            canPlay += 1
            coin -= 1
        
        if(pair1 == card2):
            canBuy += 1
        else:
            if(visited[pair1]):
                canBuy+=1
            if(visited[pair2]):
                canBuy+=1
        
        if(canPlay == 0):
            if(canBuy > 0 and coin > 1):
                canBuy -= 1
                canPlay += 1
                coin -= 2
            else:
                break
        
        round += 1
        
    return round + 1

File: 프로그래머스/3/258707. n ＋ 1 카드게임/n ＋ 1 카드게임.java
import java.util.*;

class Solution {

    public static int solution(int coin, int[] cards) {

        int n = cards.length;
        int matching = 0;
        int canPlay = 1;
        int round = 0;

        boolean[] hasCards = new boolean[n+1];
        boolean[] checked = new boolean[n+1];

        // 초기 카드들 hasCards에 저장
        // 저장시에 pair 카드가 이미 저장되어 있다면, canPlay++한다.
        int initSize = n / 3;
        for (int i = 0; i < initSize; i++) {
            int card = cards[i];
            hasCards[card] = true;

            int pair = n+1-card;
            if(hasCards[pair]) canPlay++;
        }

        // 매번 canPlay-- 해주면서 0이 되면 멈춘다.
        while(canPlay-- > 0) {
            int idx = initSize + (2 * round++);
            if(idx == n) return round;

            // 새로 나왔다는 걸 checked에 
            int fCard = cards[idx];
            int sCard = cards[idx+1];
            checked[fCard] = true;
            checked[sCard] = true;

            int fPair = n + 1 - fCard;
            int sPair = n + 1 - sCard;
            int[] pairs = new int[] {fPair, sPair};
            
            for(int pair : pairs) {
                if(coin > 0 && hasCards[pair]) {
                    // 페어 카드를 처음에 가지고 있었다면
                    coin--;
                    canPlay++;
                }
                if(checked[pair]) {
                    // 페어 카드가 처음 이후에 등장했다면 살 수 있다는 걸 체크해둔다.
                    matching++;
                }
            }
            
            // 처음에 뽑은 두 카드의 합이 n+1 이라면 matching 에서 하나를 뺀ㄴ다.
            if(fCard + sCard == n+1) matching--;

            // 매칭된 것에서 하나 빼서 라운드를 추가한다!
            if(canPlay == 0 && coin > 1 && matching > 0) {
                matching--;
                coin-=2;
                canPlay++;
            }

        }

        return round;
    }

}